test/client_reset_test.dart

```diff
@@ -0,0 +1,256 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:async';
+
+import 'package:test/test.dart' hide test, throws;
+import '../lib/realm.dart';
+import '../lib/src/configuration.dart' show ClientResetHandlerInternal, ClientResyncModeInternal, BeforeResetCallback, AfterResetCallback, ClientResetCallback;
+import 'test.dart';
+
+Future<void> main([List<String>? args]) async {
+  await setupTests(args);
+
+  baasTest("Configuration.flexibleSync set recoverOrDiscard as a default resync mode", (appConfiguration) async {
+    final app = App(appConfiguration);
+    final user = await getIntegrationUser(app);
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: ManualRecoveryHandler((syncError) {}),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.manual);
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const DiscardUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.discardLocal);
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const RecoverUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.recover);
+
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const RecoverOrDiscardUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.recoverOrDiscard);
+
+    expect(Configuration.flexibleSync(user, [Task.schema, Schedule.schema]).clientResetHandler.clientResyncMode, ClientResyncModeInternal.recoverOrDiscard);
+  });
+
+  baasTest('ManualRecoveryHandler error is reported in callback', (appConfig) async {
+    final app = App(appConfig);
+    final user = await getIntegrationUser(app);
+
+    final resetCompleter = Completer<ClientResetError>();
+    final config = Configuration.flexibleSync(
+      user,
+      [Task.schema, Schedule.schema],
+      clientResetHandler: ManualRecoveryHandler((syncError) {
+        resetCompleter.complete(syncError);
+      }),
+    );
+
+    final realm = await Realm.open(config);
+    await realm.syncSession.waitForUpload();
+
+    await triggerClientReset(realm);
+
+    final error = await resetCompleter.future;
+    expect(error.message, contains('Bad client file identifier'));
+  });
+
+  for (Type clientResetHandlerType in [
```

Comment:
These for loops are repeated several times - can we move all the parameterized tests in a single for loop?